scipy/optimize/tests/test__shuffled_complex_evolution.py

```diff
@@ -0,0 +1,334 @@
+                                            self.ub, printit=1)
+        assert_equal(result.fun, rosen(result.x))
+        result = shuffled_complex_evolution(rosen, self.x0, self.lb,
+                                            self.ub, printit=2)
+        assert_equal(result.fun, rosen(result.x))
+        result = shuffled_complex_evolution(rosen, self.x0, self.lb,
+                                            self.ub, printit=1, maxn=10)
+        assert_equal(result.fun, rosen(result.x))
+        result = shuffled_complex_evolution(rosen, self.x0, self.lb,
+                                            self.ub, printit=2, maxn=10)
+        assert_equal(result.fun, rosen(result.x))
+        result = shuffled_complex_evolution(
+            rosen, [0.5] * 5, [0.] * 5, [2.] * 5, printit=1,
+            pcento=10)
+        print(result.message)
+        assert_equal(result.fun, rosen(result.x))
+
+        # sampling
+        result = shuffled_complex_evolution(
+            rosen, self.x0, self.lb, self.ub,
+            sampling=['half-open'] * len(self.lb))
+        assert_almost_equal(result.x, self.rosenx, decimal=3)
+        smpls = ['half-open', 'right-half-open', 'left-half-open', 'open',
+                 'log']
+        for sm in smpls:
+            result = shuffled_complex_evolution(rosen, self.x0, self.lb,
+                                                self.ub, sampling=sm)
+            assert_almost_equal(result.x, self.rosenx, decimal=3)
+        lb = [-1, -1.]
+        result = shuffled_complex_evolution(rosen, self.x0, lb, self.ub,
+                                            sampling='log')
+        assert_almost_equal(result.x, self.rosenx, decimal=3)
+        lb = [0., 0.]
+        result = shuffled_complex_evolution(rosen, self.x0, lb, self.ub,
+                                            sampling='log')
+        assert_almost_equal(result.x, self.rosenx, decimal=3)
+
+        # maxit
+        result = shuffled_complex_evolution(self.negative_rosen, self.x0,
+                                            self.lb, self.ub, maxn=100,
+                                            polish=True, maxit=True)
+        assert_equal(result.fun, self.negative_rosen(result.x))
+
+    def test_errors(self):
+        # test that the bounds checking works
+        func = rosen
+        x0 = [1.]
+        bounds = [(-3)]
+        assert_raises(TypeError, shuffled_complex_evolution, func, x0, bounds)
+        bounds = [(-1, 1), (-1, 1)]
+        assert_raises(ValueError, shuffled_complex_evolution, func, x0,
+                      bounds, sampling='unknown')
+        # test correct bool string
+        assert_raises(ValueError, _strtobool, 'Ja')
+        # test no initial population found
+        func = deb03
+        x0 = [-0.5, -0.5]
+        bounds = [(-1, 0), (-1, 0.)]  # should be (0, 1) to work
+        assert_raises(ValueError, shuffled_complex_evolution, func, x0,
+                      bounds, printit=1)
```

Comment:
Please test that the correct error messages are shown using
```python
with pytest.raises(ValueError, match="error messsage")
``` 

Other necessary tests include objectives that:

- return `NaN`
- return `inf`